### PR TITLE
Sync with NoahMP Github version with all NoahMP updates since v4.3

### PR DESCRIPTION
**Pull request by cenlinhe**
TYPE: new feature

KEYWORDS: NoahMP

SOURCE: Cenlin He (NCAR)

DESCRIPTION OF CHANGES:
This is step one to synchronize with latest NoahMP Github code version in order to prepare for connecting WRF directly to NoahMP unified Github repo. 

Specifically, the updates in NoahMP since version 4.3:
1. three additional runoff schemes added (OPT_RUN=6, 7, 8)
2. new tile drainage capability/scheme added (OPT_TDRN)
3. New soil time step control capability added to allow users specify a different soil time step from main 
NoahMP time step
4. additional water and energy budget terms added for output and diagnosis
5. updated flood irrigation parameters
6. add constraints for snow thickness during compaction
7. updated snow viscosity coefficient in compaction process
8. add canopy heat storage in canopy temperature calculation
9. updated formulation and parameter values for TOPMODEL groundwater scheme (OPT_RUN=1)
10. Bug fixes: SWE cap value in initialization, NoahMP snow depth update, snow layer index update, floating 
issues in runoff scheme, and BATS albedo scheme

LIST OF MODIFIED FILES: list of changed files (use `git diff --name-status master` to get formatted list)
M       Registry/Registry.EM_COMMON
M       Registry/registry.noahmp
M       dyn_em/module_first_rk_step_part1.F
M       dyn_em/start_em.F
M       phys/module_diag_misc.F
M       phys/module_diagnostics_driver.F
M       phys/module_physics_init.F
M       phys/module_sf_noahmp_glacier.F
M       phys/module_sf_noahmp_groundwater.F
M       phys/module_sf_noahmpdrv.F
M       phys/module_sf_noahmplsm.F
M       phys/module_surface_driver.F
M       run/MPTABLE.TBL
M       run/README.namelist
M       run/SOILPARM.TBL

TESTS CONDUCTED: 
1. successfully compiled and tested for running in NCAR supercomputer by Cenlin He (NCAR)
2. Jenkins tests are all pass.

RELEASE NOTE: Updates from NoahMP and linking WRF directly to NoahMP Github repo.

### CODE REVIEW

File: Registry/registry.noahmp [hunk 1]
```diff
@@ -79,27 +79,139 @@ state    real   SMOISEQ    ilj     -       1      Z     r     "SMOISEQ"    "EQ.
...
+state    real   eflxbxy       ij      -       1      -     rd=(interp_mask_land_field:lu_index)u=(copy_fcnm)       "eflxbxy "      "bottom soil heat flux"                "W/m2"
+state    real   acc_dwaterxy  ij      -       1      -     rd=(interp_mask_land_field:lu_index)u=(copy_fcnm)       "acc_dwater"    "accumulated canopy,snow,soil water change between dt_soil"   "mm"
+state    real   acc_prcpxy    ij      -       1      -     rd=(interp_mask_land_field:lu_index)u=(copy_fcnm)       "acc_prcp"    "accumulated precipitation between dt_soil"   "mm"
+state    real   acc_ecanxy    ij      -       1      -     rd=(interp_mask_land_field:lu_index)u=(copy_fcnm)       "acc_ecan"    "accumulated net canopy evaporation between dt_soil"   "mm"
+state    real   acc_etranxy   ij      -       1      -     rd=(interp_mask_land_field:lu_index)u=(copy_fcnm)       "acc_etran"   "accumulated transpiration between dt_soil"   "mm"
+state    real   acc_edirxy    ij      -       1      -     rd=(interp_mask_land_field:lu_index)u=(copy_fcnm)       "acc_edir"    "accumulated net soil evaporation between dt_soil"   "mm"
```
Comment (weiwangncar) [inline Registry/registry.noahmp:212]: @cenlinhe Are these new history output variables packaged?

**Comment by cenlinhe** [comment]
I just found a bug in defining an "inout" variable and am re-doing all the tests for the bug fix. I will update this PR soon after successful test.

File: run/README.namelist [hunk 2]
```diff
@@ -1278,37 +1285,45 @@ Options for use with the Noah-MP Land Surface Model:
...
+                                                     2 = micro/drip irrigation
+                                                     3 = surface flooding  
+ opt_tdrn                           = 0,        ! Noah-MP tile drainage option (currently only tested and works with opt_run=3)
+                                                     geogrid must have been run with GEOGRID.TBL.ARW.noahmp, use with caution
+                                                     0 = No tile drainage
+                                                     1 = Simple drainage
+                                                     2 = Hooghoudt's equation based tile drainage
+ soiltstep                          = 0.0,      ! Noah-MP soil process timestep (seconds) for solving soil water and temperature
+                                                     0 = default, the same as main NoahMP model timestep
+                                                     N * dt_noahmp, typically 15min or 30min
```
Comment (weiwangncar) [inline run/README.namelist:1326]: @cenlinhe Can you add noahmp_output, noahmp_acc_dt, and possibly opt_infdv, opt_tdrn to this file?

**Comment by cenlinhe** [comment]
The PR is now updated. I fixed the bugs and add info to README.namelist.

**Comment by davegill** [comment]
@weiwangncar @dudhia 
Folks,
If people are OK with these mods (and really, these are in the NoahMP repo, so we are basically *given* these AS-IS), we should approve to merge in the next step as quickly as possible.

**Review by weiwangncar — APPROVED** [review]

**Comment by dudhia** [comment]
I could approve but some of these mods are mine. Dave could go ahead and do the software approve.

**Review by davegill — APPROVED** [review]

**Pull request merged**